llama_index/query_engine/retry_query_engine.py

```diff
@@ -0,0 +1,292 @@
+                raise ValueError(f"Unexpected response type {type(response)}")
+
+            response_eval = evaluator.evaluate(query_bundle.query_str, typed_response)
+            if response_eval == "YES":
+                return response
+            else:
+                logger.debug(
+                    f">Binary evaluator returned NO, response: \
+                    {typed_response.response}"
+                )
+
+            old_source_nodes = evaluator.get_context(typed_response)
+            new_index = GPTListIndex.from_documents(old_source_nodes)
+            if self.check_source:
+                source_node_evals = evaluator.evaluate_source_nodes(
+                    query_bundle.query_str, typed_response
+                )
+                assert len(old_source_nodes) == len(source_node_evals)
+                new_source_docs: List[Document] = []
+                for i in range(len(source_node_evals)):
+                    if eval == "YES":
+                        new_source_docs.append(old_source_nodes[i])
+                    else:
+                        logger.debug(
+                            f">Source evaluator return NO, node: {old_source_nodes[i]}"
+                        )
+                if len(new_source_docs) == 0:
+                    # TODO: Add context-less query.
+                    logger.warn("No source nodes are relevant")
+                if new_source_docs:
+                    new_index = GPTListIndex.from_documents(new_source_docs)
+
+            new_retriever = new_index.as_retriever(
+                optimizer=SentenceEmbeddingOptimizer(
+                    percentile_cutoff=self.percentile_cutoff
+                )
+            )
+
+            if self.use_shrinking_percentile_cutoff:
+                new_percentile_cutoff = self.percentile_cutoff - 0.1
+            text_qa_template = QuestionAnswerPrompt(
+                self.get_error_correcting_qa_tmpl(typed_response.response)
+            )
+            refine_template = RefinePrompt(
+                self.get_error_correcting_refine_tmpl(typed_response.response)
+            )
+            new_query_engine = RetryQueryEngine.from_args(
+                text_qa_template=text_qa_template,
+                refine_template=refine_template,
+                retriever=new_retriever,
+                check_source=self.check_source,
+                percentile_cutoff=new_percentile_cutoff,
+                use_shrinking_percentile_cutoff=self.use_shrinking_percentile_cutoff,
+                max_retries=self.max_retries - 1,
+            )
+            logger.debug(f">Retrying query. Retries left: {self.max_retries - 1}")
+            response = new_query_engine.query(query_bundle)
+
+        return response
+
+    async def _aquery(self, query_bundle: QueryBundle) -> RESPONSE_TYPE:
+        """Not supported."""
+        return self._query(query_bundle)
+
+    def get_error_correcting_qa_tmpl(self, neg_ex: Optional[str]) -> str:
```

Comment:
So this is interesting. Your approach is adding a negative answer example to the prompt itself instead in the query. Did you think about the latter approach as well? (the latter would be amenable to the design of just RetryQueryEngine against a BaseQueryEngine without any visibility in the underlying abstractions)